### PR TITLE
database tables defined in model.py

### PR DESCRIPTION
**Pull request by spacemansteve**
and use sqlalchemy orm api.

### CODE REVIEW

**Comment by romanchyla** [comment]
In my experience, the ORM layer can add a LOT of overhead; so while I see the rationale and have nothing against the code, it is important to run some performance comparison. Could you run it on few million records please?

**Comment by spacemansteve** [comment]
I tested performance using a million bibcodes on adswhy.  The machine's fluctuating load complicates interpreting the results.  Ingest of nonbib column files uses psycogp copy_from so it is not affected by the sqlalchemy api.  It took 4 minutes to read the column files for 1 million bibcodes and perform a sql join across the 13 column file.  I expected something closer to 2.5 minutes and attribute the difference to reduced I/O performance due (fulltext ingest was also running).  Computing ingest values for the 1 million bibcodes took 24 minutes.  This operation is sensitive to the sqlalchemy api.  I expected something closer to half that time, perhaps much of the discrepancy is again from system load.  I will run further tests.

**Comment by spacemansteve** [comment]
Changed to support incremental reads via sqlalchemy's yield_per

File: run.py [hunk 1]
```diff
@@ -37,55 +39,71 @@ def load_column_files(config, sql_sync):
...
+    session = Session()
+    session.execute('set search_path to {}'.format(schema))
     tmp = []
-    for current_delta in results:
-        row = nonbib.get_row_view(current_delta['bibcode'])
-        rec = NonBibRecord(**dict(row))
+    i = 0
+    max_rows = config['MAX_ROWS']
+    for current_delta in session.query(models.DeltaTable).yield_per(100):
+        row = nonbib.get_row_view(current_delta.bibcode)
```
Comment (romanchyla) [inline run.py]: tihs code is missing, also - can you try to find a better name for it? it is not clear what it's supposed to do (and on what)

i'm also seeing it used as a classmethod and as in this instance as a function (of a module); there is some confusion

File: run.py [hunk 2]
```diff
@@ -100,50 +118,51 @@ def nonbib_delta_to_master_pipeline(schema, batch_size=1):
...
-
-def metrics_delta_to_master_pipeline(metrics_schema, nonbib_schema, batch_size=1):
+def metrics_delta_to_master_pipeline(metrics_engine, metrics_schema, nonbib_schema, batch_size=1):
     """send data for changed metrics to master pipeline
 
     the delta table was computed by comparing to sets of nonbib data
     perhaps ingested on succesive days"""
     m = metrics.Metrics(metrics_schema)
-    nonbib = row_view.SqlSync(nonbib_schema)
+    nonbib = nonbib.NonBib(nonbib_schema)
```
Comment (romanchyla) [inline run.py]: so this is confusing; it overrides module (in the function scope) with the object

File: run.py (hunk 1 above)
Comment (spacemansteve) [reply run.py]: get_row_view renamed to get_by_bibcode
https://github.com/spacemansteve/AdsDataSqlSync/blob/master/adsdata/nonbib.py#L197

File: run.py [hunk 3]
```diff
@@ -240,140 +259,114 @@ def main():
...
         m = metrics.Metrics(args.metricsSchemaName)
-        m.drop_metrics_table()
+        m.drop_metrics_table(metrics_db_engine)
+
     elif args.command == 'populateMetricsTable' and args.rowViewSchemaName and args.metricsSchemaName:
-        # m = metrics.Metrics(None, {'FROM_SCRATCH': False, 'COPY_FROM_PROGRAM': False})
-        m = metrics.Metrics(None, {'FROM_SCRATCH': True, 'COPY_FROM_PROGRAM': False})
-        m.update_metrics_all(args.rowViewSchemaName)
+        m = metrics.Metrics()
+        m.update_metrics_all(metrics_db_conn, nonbib_db_conn, args.rowViewSchemaName)
```
Comment (romanchyla) [inline run.py]: this call also loads all results into memory before copying them; please change all occurences where a large datasets are being loaded into memory

also, I'm seeing the following being repeated on many occassions

```
Session = sessionmaker()
sess = Session(bind=db_conn)
```

you are passing the db connections to methods of a metrics/nonbib instance; consider initializing these instances with the db connections (i.e. let them establish connection and keep the session) - that way perhaps many lines can be dropped fromthe code

File: run.py [hunk 4]
```diff
@@ -240,140 +259,114 @@ def main():
...
     elif args.command == 'renameSchema' and args.rowViewSchemaName and args.rowViewBaselineSchemaName:
-        sql_sync = row_view.SqlSync(args.rowViewSchemaName, config)
-        sql_sync.rename_schema(args.rowViewBaselineSchemaName)
+        sql_sync = nonbib.NonBib(args.rowViewSchemaName)
+        sql_sync.rename_schema(nonbib_db_conn, args.rowViewBaselineSchemaName)
+
     elif args.command == 'createDeltaRows' and args.rowViewSchemaName and args.rowViewBaselineSchemaName:
-        sql_sync = row_view.SqlSync(args.rowViewSchemaName, config)
-        sql_sync.create_delta_rows(args.rowViewBaselineSchemaName)
+        sql_sync = nonbib.NonBib(args.rowViewSchemaName)
```
Comment (romanchyla) [inline run.py]: the sql_sync instance should be created before the block of if statements

File: run.py (hunk 3 above)
Comment (spacemansteve) [reply run.py]: all results are no longer loaded into memory: https://github.com/spacemansteve/AdsDataSqlSync/blob/master/adsdata/metrics.py#L177

**Comment by romanchyla** [comment]
i've put the comment on the wrong line, see this:
https://github.com/spacemansteve/AdsDataSqlSync/blob/b1b2ebe30447c9ea1008d727d9dfa8ed4126de07/adsdata/metrics.py#L116

my other comments are still valid

File: run.py (hunk 4 above)
Comment (spacemansteve) [reply run.py]: fixed in new commit: https://github.com/spacemansteve/AdsDataSqlSync/commit/3031125f3156da97f2176dc67af5734abb378660

File: run.py (hunk 2 above)
Comment (spacemansteve) [reply run.py]: addressed in new commit: https://github.com/spacemansteve/AdsDataSqlSync/commit/04a256b429c48c444a8984d69487f50ee6aff8b2

File: run.py (hunk 3 above)
Comment (spacemansteve) [reply run.py]: addressed in new commit: https://github.com/spacemansteve/AdsDataSqlSync/commit/04a256b429c48c444a8984d69487f50ee6aff8b2

Comment (spacemansteve) [reply run.py]: regarding db connections, I decided to have the calling function pass in the connection so it is clear the calling function was also responsible for closing the connection.  I will investigate best practices and existing project code to decide if/how to change.

**Comment by romanchyla** [comment]
it is not immediately clear how this affectsspeed - but i think the ORM is used on every row that gets loaded from metrics; it is the met.get_row_view(); 

can you confirm that the change for ORM doesn't affect the execution speed dramatically?

**Comment by spacemansteve** [comment]
I can confirm that performance is acceptable with the ORM change.

**Pull request merged**